### PR TITLE
+Cleanup MOM_set_diffusivity and MOM_geothermal

### PR DESCRIPTION
**Pull request by Hallberg-NOAA**
Refactored MOM_set_diffusivity.F90, MOM_geothermal.F90, MOM_diabatic_driver.F90
and related modules to improve the code structure, and to avoid unnecessarily setting
or using 3-d arrays that are not used themselves.  This includes the elimination
of at least 3 3-d arrays and a dramatic simplification of the geothermal heating
routine in ALE mode.  The registration calls for some modules and diagnostics
were changed, so there are changes in the order of the entries in many
MOM_parameter_doc files.  Also added the new runtime parameter PRANDTL_EPBL, and
corrected two diagnostics of the thickness and temperature changes due to
geothermal heating that were just wrong before. All tested solutions are bitwise
identical, but the corrections to diagnostics cause some automated regression
tests fail.  The commits in this PR include:

- NOAA/GFDL/MOM6@94a833839 Fixed the diagnostic internal_heat_temp_tendency
- NOAA/GFDL/MOM6@cd3ebe41f Merge branch 'dev/gfdl' into set_diffusivity_cleanup
- NOAA/GFDL/MOM6@5df01372d +Added the runtime parameter PRANDTL_EPBL
- NOAA/GFDL/MOM6@fc806af64 Reordering of code in calculate_bkgnd_mixing
- NOAA/GFDL/MOM6@2968e4dfe Refactored diagnostics in geothermal_in_place
- NOAA/GFDL/MOM6@c4e145519 Eliminated unused variables in diabatic_ALE
- NOAA/GFDL/MOM6@c61232f06 +Overloaded the interface to geothermal
- NOAA/GFDL/MOM6@986bfbd24 Reuse auxiliary salinity array in set_diffusivity
- NOAA/GFDL/MOM6@ca3a5a988 +Initialize tidal_mixing from set_diffusivity_init
- NOAA/GFDL/MOM6@03c7433a9 Removed Kd_lay from diabatic_ALE routines
- NOAA/GFDL/MOM6@741b9c703 +Work with 2-d slice of Kd_int in set_diffusivity
- NOAA/GFDL/MOM6@f05782194 +Made Kd_lay argument to set_diffusivity optional
- NOAA/GFDL/MOM6@8d1539e8a +Made elements of tidal_mixing_CS private
- NOAA/GFDL/MOM6@c007a1395 +Move diagnostics out of MOM_CVMix_ddiff
- NOAA/GFDL/MOM6@8ef820ea1 +Remove 3-d diagnostic arrays in MOM_bkgnd_mixing
- NOAA/GFDL/MOM6@db4d3f0f8 White space cleanup in MOM_generic_tracer
- NOAA/GFDL/MOM6@902f7a9f2 +Made Kd_lay arg to calculate_tidal_mixing optional
- NOAA/GFDL/MOM6@449d568fe +Eliminated sfc_bkgnd_mixing

### CODE REVIEW

**Comment by Hallberg-NOAA** [comment]
This is a complicated PR. It will pass the pipeline regression tests apart form changes in the MOM_parameter_doc files, but as expected the Travis CI tests are failing in the regression checks of the diagnostics for three separate reasons:

1. The order of some diagnostics have changed for reasons I do not fully appreciate, but perhaps because of changes on the order with which they are registered, or perhaps the order in which they are posted.
2. Two diagnostics in geothermal (of internal heat driven tendencies) that were fundamentally wrong before have been corrected.
3. One of these geothermal diagnostics that was wrong and a second related diagnostic have been refactored for efficiency and reduced memory usage in a way that changes these diagnostics at roundoff.

  If this PR is too complicated to evaluate, the various commits could be broken up into multiple stages:

1. Fix the fundamentally broken geothermal diagnostics (Will fail diagnostic regressions in TC, but would be ~20 lines of code with no changes to MOM6-examples)
2. White space, comment and similar trivial changes in MOM_generic_tracer and MOM_set_diffusivity (Green button push)
3. Add PRANDTL_EPBL parameter and move initialization of MOM_tidal mixing into set_diffusivity_init (Might fail diagnostic regressions in TC due to reordered but identical diagnostics and will change MOM_parameter_doc files)
4. Refactoring of geothermal into an overloaded interface (Green button push)
5. Optimize 2 geothermal diagnostics when in ALE mode for efficiency (Will fail diagnostic regressions in TC, but would be ~20 lines of code with no changes to MOM6-examples)
6. Make the series of changes in MOM_set_diffusivity to make Kd_lay an optional argument. (Lots of lines of code in multipe files, but probably passing TC regressions leading to a green button push)

  Please note that several of these changes would have to be carried out in sequence if the testing and approval is to be simplified, and that some of these PRs can not even be requested until the predecessor has been approved.  Note also that many of these PRs would occur in the opposite of the order in which they were developed, so replacing this one PR as a series of 6 PRs would require all of the code changes to be redone. However, if the prospective reviewers are finding this to complicated an onerous to review, and if one of the other people with write access to dev/gfdl were willing to commit to handling PRs 1-5 in this list promptly, I would be willing to to withdraw this large PR and redo it in the 6 PR sequence described above.

**Comment by gmacgilchrist** [comment]
Concerning the `MOM_geothermal.F90` diagnostic calculation changes [here](https://github.com/NOAA-GFDL/MOM6/pull/1228/commits/94a8338399cae10ec97418cf5f427077e54ddd6e), these look correct to me. I had previously noted that these diags were coming out zero, but had not got round to investigating.

**Review by ashao — COMMENTED** [review]
Have to run now, but I'll finish the review later, including overall comments.

File: src/parameterizations/vertical/MOM_bkgnd_mixing.F90
Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_bkgnd_mixing.F90:341]: Curiously, if you are willing to work in radians instead of degrees, this value of the minimum value of sin(lat) is also the minimum value of lat to within 64-bit machine precision, since `sin(x) = x * (1 - x^2/3! + x^4/5! - ...)` and (1.e-10)**2/6 < 1.0e-17. So maybe we just need to revise the description in the comment (or not).

File: src/parameterizations/vertical/MOM_geothermal.F90
Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_geothermal.F90:163]: It is necessary to set `work3d` to 0 here if this heat_tendency diagnostic is in use, for precisely the same reason why the code storing h_old and T_old needed to be moved outside of the loops _with conditional exits_ where the heating occurs.

  I strongly agree that shorter diagnostic id names here would be less cumbersome.

File: src/parameterizations/vertical/MOM_bkgnd_mixing.F90
Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_bkgnd_mixing.F90:314]: Thank you for pointing this out. I have another bug-fix to MOM_bkgnd_mixing.F90 that will follow as soon as this PR is merged in, and I will fix this typo then.

File: src/parameterizations/vertical/MOM_diabatic_driver.F90
Comment (adcroft) [inline src/parameterizations/vertical/MOM_diabatic_driver.F90:832]: This is a bug fix that doesn't change answers for the M6E suite but does affect OM4.1 for which Pr=1.25 . We can forego a run-time parameter so long as no one else has a non-unity Pr.

**Review by adcroft — CHANGES_REQUESTED** [review]
This is overall a good tidy up.

The new `geothermal_in_place()` is a much cleaner implementation for when we are in general coordinate mode. Stripping out the interleaving of entrainment calculations is both more efficient and makes the code easier to understand. It is a shame the reverse can't be done for `isopycnal_geothermal()`.

The one change I request: I do not think this is useful example of overloaded interfaces so please call `geothermal_in_place()` and `isopycnal_geothermal()` directly. When following the call tree it is better to know where you are going. I've spent too much of my life reading the wrong version of a routine.

File: src/parameterizations/vertical/MOM_diabatic_driver.F90
Comment (ashao) [inline src/parameterizations/vertical/MOM_diabatic_driver.F90:957]: Since `eatr(k+1)=ebtr(k)` in ALE mode, this diagnostic should always be 0. Maybe don't register in ALE mode?

**Review by ashao — CHANGES_REQUESTED** [review]
Looks good to me and definitely clears up the code flow quite a bit. I'm not sure why the diagnostic tests are failing either.

File: src/parameterizations/vertical/MOM_CVMix_ddiff.F90
Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_CVMix_ddiff.F90:150]: I suspect that we might have detected this if the compilers had not be treating this array as intent(inout).

File: src/parameterizations/vertical/MOM_diabatic_driver.F90
Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_diabatic_driver.F90:833]: I introduced a new runtime parameter (PRANDTL_EPBL) that is used set ePBL_Prandtl with a default value of 1 so this can't possibly affect OM4.1.  We can consider using a global Prandtl number to set the defaults throughout the code, but this PR does not do this.

File: src/parameterizations/vertical/MOM_bkgnd_mixing.F90
Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_bkgnd_mixing.F90]: This typo has been corrected along with some other requested changes.

File: src/parameterizations/vertical/MOM_CVMix_ddiff.F90
Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_CVMix_ddiff.F90:142]: The declaration statements have been reordered to match the order of the arguments, to the extent that the Fortran standard and the use of elements of G and GV to declare array sizes permit.

File: src/parameterizations/vertical/MOM_CVMix_ddiff.F90
Comment (Hallberg-NOAA) [reply src/parameterizations/vertical/MOM_CVMix_ddiff.F90]: The overloaded interface to geothermal() has been replaced with separate interfaces to geothermal_entraining() and geothermal_in_place().

**Review by adcroft — APPROVED** [review]
https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/11448

**Pull request merged**
